automap.c

```diff
@@ -832,78 +1814,231 @@ fam_get(FAMObject *self, PyObject *args)
 static PyObject *
 fam_items(FAMObject *self)
 {
-    return view(self, ITEMS);
+    return famv_new(self, ITEMS);
 }
 
 
 static PyObject *
 fam_keys(FAMObject *self)
 {
-    return view(self, KEYS);
+    return famv_new(self, KEYS);
 }
 
 
 static PyObject *
 fam_values(FAMObject *self)
 {
-    return view(self, VALUES);
+    return famv_new(self, VALUES);
 }
 
 
-static PyMethodDef fam_methods[] = {
-    {"__getnewargs__", (PyCFunction) fam___getnewargs__, METH_NOARGS, NULL},
-    {"__reversed__", (PyCFunction) fam___reversed__, METH_NOARGS, NULL},
-    {"__sizeof__", (PyCFunction) fam___sizeof__, METH_NOARGS, NULL},
-    {"get", (PyCFunction) fam_get, METH_VARARGS, NULL},
-    {"items", (PyCFunction) fam_items, METH_NOARGS, NULL},
-    {"keys", (PyCFunction) fam_keys, METH_NOARGS, NULL},
-    {"values", (PyCFunction) fam_values, METH_NOARGS, NULL},
-    {NULL},
-};
-
+// This macro can be used with integer and floating point NumPy types, given an `npy_type` and a specialized `insert_func`. Uses context of `fam_init` to get `fam`, `contiguous`, `a`, `keys_size`, and `i`. An optional `pre_insert` function can be supplied to transform extracted values before calling the insert function.
+# define INSERT_SCALARS(npy_type, insert_func, pre_insert)    \
+if (contiguous) {                                             \
+    npy_type* b = (npy_type*)PyArray_DATA(a);                 \
+    npy_type* b_end = b + keys_size;                          \
+    while (b < b_end) {                                       \
+        if (insert_func(fam, pre_insert(*b), i, -1)) {        \
```

Comment:
`pre_insert(*b)` -> `(*b)` when `pre_insert` is empty is really clever

Reply:
Thanks! I was worried that the macro usage here would fall apart with the one-off need for converter function, but being able to use empty macro arguments is pretty nice!